### PR TITLE
jewel: rbd: object-map: batch updates during trim operation

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/17843

### CODE REVIEW

File: src/librbd/AioObjectRequest.h [hunk 1]
```diff
@@ -331,11 +331,14 @@ class AioObjectRemove : public AbstractAioObjectWrite {
 
 class AioObjectTrim : public AbstractAioObjectWrite {
 public:
+  // we'd need to only conditionally specify if a post object map
```
Comment (amitkumar50) [inline src/librbd/AioObjectRequest.h]: @smithfarm 
I believe multi-line comment should be inside /* */
Please replace we'd with we would

**Comment by smithfarm** [comment]
@amitkumar50 Backports are cherry-picked from master, so any changes would have to go into master, first.

Regarding contractions, see the discussion in #16705

**Comment by smithfarm** [comment]
@dillaman @trociny This PR was included in this run, which is exhibiting a new failure: http://pulpito.ceph.com/smithfarm-2017-08-24_15:58:59-rbd-wip-jewel-backports-distro-basic-smithi/

**Comment by trociny** [comment]
All tests failed for ceph_test_librbd_fsx like below:
```
2017-08-25T03:23:10.028 INFO:teuthology.orchestra.run.smithi193.stdout:428 clone        6 order 18 su 1024 sc 6
2017-08-25T03:23:10.871 INFO:teuthology.orchestra.run.smithi193.stdout:truncating image image_client.0-clone5 from 0x37660d2 (overlap 0x47dc2d) to 0x23ee16
2017-08-25T03:23:11.063 INFO:teuthology.orchestra.run.smithi193.stdout:rbd_get_flags() indicates object map is invalid
2017-08-25T03:23:11.103 INFO:teuthology.orchestra.run.smithi193.stdout:do_clone: ops->resize: Invalid argument

When I revert this PR in wip-jewel-backports branch the tests start to pass locally. So the failures are most likely related to this PR.

**Comment by trociny** [comment]
@smithfarm The problem with this PR is that in the master this "batch update" change [1] was merged before "order concurrent updates" [2], while in jewel "order concurrent updates" is already backported [3]. Due to vice verse order, when backporting "order concurrent updates" Jason skipped changes in "TrimRequest.cc" that should be applied to `send_pre_copyup` and `send_post_copyup` function (introduced by "batch update" change). So your cherry-pick miss this changes too and you need to add them manually.

Do you want to work on this? Or I can do this if you wish.

[1] https://github.com/ceph/ceph/pull/11510
[2] https://github.com/ceph/ceph/pull/12420
[3] https://github.com/ceph/ceph/pull/12909
[4] https://github.com/ceph/ceph/pull/12909/commits/c53df3780dd9221cfe602c09651eeee06046ebeb

**Comment by smithfarm** [comment]
@trociny Thanks for debugging this. I think you have given me enough hints that I'll be able to track down and apply the needed changes myself. I'll ping you for review if/when that happens.

**Comment by smithfarm** [comment]
@trociny Thanks for debugging this. Thanks to your analysis, I think I was able to track down and apply the needed changes myself (see the new commit).

**Comment by smithfarm** [comment]
The last force-push just tweaks the commit message.

**Comment by smithfarm** [comment]
test this please

**Comment by smithfarm** [comment]
@dillaman This passed an rbd suite at http://tracker.ceph.com/issues/20613#note-31 (there was 1 failure that passed on rerun).

The rados run is putting up a bit of a fight; there are a couple failures but they don't appear to be reliably reproducible or related to rbd: http://tracker.ceph.com/issues/20613#note-25

I also ran three different upgrade suites and all the failures were either not reproducible or were "hand-waved away". 

Please review.

**Pull request merged**